src/mds/Migrator.cc

```diff
@@ -525,22 +525,13 @@ void Migrator::handle_mds_failure_or_stop(mds_rank_t who)
 	break;
 
       case IMPORT_LOGGINGSTART:
+      case IMPORT_ACKING:
 	ceph_assert(dir);
-	dout(10) << "import state=loggingstart : reversing import on " << *dir << dendl;
+	dout(10) << "import state=" << get_import_statename(q->second.state)
+		 << " : reversing import on " << *dir << dendl;
 	import_reverse(dir);
 	break;
 
-      case IMPORT_ACKING:
-	ceph_assert(dir);
-	// hrm.  make this an ambiguous import, and wait for exporter recovery to disambiguate
```

Comment:
this line says that the export will do the disambiguation. You mentioned "should reverse the process to clear the importing state if the exporter restarts at this time, but it doesn't do that actually". So an exporter once restarted, doesn't perform recovery and thus no disambiguation, is it?

Reply:
> this line says that the export will do the disambiguation. You mentioned "should reverse the process to clear the importing state if the exporter restarts at this time, but it doesn't do that actually". So an exporter once restarted, doesn't perform recovery and thus no disambiguation, is it?

Yes, it is. `import_reverse` can reverse the process when the state is `IMPORT_ACKING`. But `import_reverse` has a bug and we need to fix it.

Reply:
If that is the case then I feel rather than changing the way we handle the cases, we should address the bug in `import_reverse`.

Reply:
> If that is the case then I feel rather than changing the way we handle the cases, we should address the bug in `import_reverse`.

The bug in `import_reverse` is that it contains the code to handle state `IMPORT_ACKING` but it will never be executed because the state is modified to `IMPORT_ABORTING` at the beginning. Move `stat.state = IMPORT_ABORTING` to the end of `import_reverse` so that it can handle the state `IMPORT_ACKING`. And then we call `import_reverse` when the state is `IMPORT_ACKING` and the process needs to be reversed.